core/cache/charmconfigwatcher.go

```diff
@@ -0,0 +1,279 @@
+
+	masterSettings map[string]interface{}
+	branchDeltas   settings.ItemChanges
+	configHash     string
+}
+
+// newUnitConfigWatcher returns a new watcher for the input config keys
+// with a baseline hash of their config values from the input hash cache.
+// As per the cache requirements, hashes are only generated from sorted keys.
+func newCharmConfigWatcher(cfg charmConfigWatcherConfig) (*CharmConfigWatcher, error) {
+	w := &CharmConfigWatcher{
+		notifyWatcherBase: newNotifyWatcherBase(),
+		initComplete:      make(chan struct{}),
+		unitName:          cfg.unitName,
+		appName:           cfg.appName,
+	}
+
+	deregister := cfg.res.registerWorker(w)
+
+	multi := cfg.hub.NewMultiplexer()
+	multi.Add(cfg.appConfigChangeTopic, w.appConfigChanged)
+	multi.Add(cfg.branchChangeTopic, w.branchChanged)
+	multi.Add(cfg.branchDeleteTopic, w.branchDeleted)
+
+	w.tomb.Go(func() error {
+		<-w.tomb.Dying()
+		multi.Unsubscribe()
+		deregister()
+		return nil
+	})
+
+	if err := w.init(cfg.model); err != nil {
+		_ = w.Stop()
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// init determines baseline master configuration, branch tracking settings
+// and the configuration hash for the watcher's unit.
+// It then closes the init channel to indicate the watcher is operational.
+func (w *CharmConfigWatcher) init(model charmConfigModel) error {
+	app, err := model.Application(w.appName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	w.masterSettings = app.Config()
+
+	for _, b := range model.Branches() {
+		unitNames := b.AssignedUnits()[w.appName]
+		if len(unitNames) == 0 {
+			continue
+		}
+
+		// A unit can only track one branch at a time,
+		// so if we find one, we are done.
+		if set.NewStrings(unitNames...).Contains(w.unitName) {
+			w.branchName = b.Name()
+			w.branchDeltas = b.Config()[w.appName]
+			break
+		}
+	}
+
+	if _, err := w.setConfigHash(); err != nil {
+		return errors.Trace(err)
+	}
+
+	close(w.initComplete)
+	return nil
+}
+
+// appConfigChanged is called when a message is received indicating changed
+// application master charm configuration.
+func (w *CharmConfigWatcher) appConfigChanged(_ string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	hashCache, ok := msg.(*hashCache)
+	if !ok {
+		logger.Errorf("programming error; application config message was not of expected type, *hashCache")
+		return
+	}
+
+	w.masterSettings = hashCache.config
+	w.checkConfig()
+}
+
+// branchChanged is called when we receive a message to say that a branch has
+// been updated in the cache.
+// If it is the branch that this watcher's unit is tracking,
+// check if the latest config delta warrants a notification.
+func (w *CharmConfigWatcher) branchChanged(_ string, msg interface{}) {
+	if !w.waitInitOrDying() {
+		return
+	}
+
+	b, okUnit := msg.(Branch)
+	if !okUnit {
+		logger.Errorf("programming error; branch change message was not of expected type, Branch")
+		return
+	}
+
+	if w.branchName != "" {
```

Comment:
Minor suggestion: you could make this block a bit shorter by examining the `if w.branchName == ""` branch first and then adding 
```
else if w.branchName != b.Name() {
  return 
}
```